src/mongoose_graphql.erl

```diff
@@ -0,0 +1,139 @@
+%% @doc This module provides main interface to graphql. It initializes schemas
+%% and allows executing queries with permissions checks.
+%% @end
+-module(mongoose_graphql).
+
+-include_lib("kernel/include/logger.hrl").
+
+%API
+-export([init/0,
+         get_endpoint/1,
+         create_endpoint/3,
+         execute/2,
+         execute/3]).
+
+-ignore_xref([create_endpoint/3]).
+
+-type request() :: #{document := binary(),
+                     operation_name := binary() | undefined,
+                     vars := map(),
+                     authorized := boolean(),
+                     ctx := map()}.
+
+-export_type([request/0]).
+
+-define(USER_EP_NAME, user_schema_ep).
+-define(ADMIN_EP_NAME, admin_schema_ep).
+
+%% @doc Create and initialize endpoints for user and admin.
+-spec init() -> ok.
+init() ->
+    create_endpoint(?USER_EP_NAME, user_mapping_rules(), schema_pattern("user")),
+    create_endpoint(?ADMIN_EP_NAME, admin_mapping_rules(), schema_pattern("admin")),
+    ok.
+
+%% @doc Get endpoint_context for passed endpoint name.
+-spec get_endpoint(atom()) -> graphql:endpoint_context().
+get_endpoint(admin) ->
+    graphql_schema:get_endpoint_ctx(?ADMIN_EP_NAME);
+get_endpoint(user) ->
+    graphql_schema:get_endpoint_ctx(?USER_EP_NAME);
+get_endpoint(Name) ->
+    graphql_schema:get_endpoint_ctx(Name).
+
+%% @doc Create a new endpoint and load schema.
+-spec create_endpoint(atom(), map(), file:filename_all()) -> gen:start_ret().
+create_endpoint(Name, Mapping, Pattern) ->
+    Res = graphql_schema:start_link(Name),
+    Ep = graphql_schema:get_endpoint_ctx(Name),
+    {ok, SchemaData} = load_multiple_file_schema(Pattern),
+    ok = graphql:load_schema(Ep, Mapping, SchemaData),
+    ok = graphql:validate_schema(Ep),
+    Res.
+
+%% @doc Execute request on a given endpoint.
+-spec execute(graphql:endpoint_context(), request()) ->
+    {ok, map()} | {error, term()}.
+execute(Ep, #{document := Doc,
+              operation_name := OpName,
+              authorized := AuthStatus,
+              vars := Vars,
+              ctx := Ctx}) ->
+    try
+        {ok, Ast} = graphql_parse(Doc),
```

Comment:
A bit strange things take place here. Spec say that `graphql:validate` can return `ok | {error, term()}` but it only calls `graphql_validate:x`, that returns only `ok` and throws errors. So that type specification should be fixed in the GraphQL library. I think all possible library failures are handled correctly here.

It would be nice to introduce such a concept here. Maybe in one of the future PRs?